### PR TITLE
feat: Add checkbox item type to button dropdown component

### PR DESCRIPTION
**Pull request by akoreman**
### Description

Currently, the ButtonDropdown component supports two types of items in this list, action items and link items. Action items allow users to invoke a callback on the item press and the link item takes the user to a URL. 

This change adds a third item type which allows the users to toggle a setting on and off. When clicked it invokes the same callback as the base items with the checkbox state as an extra argument.

Related links, issue #, if available: n/a

### How has this been tested?

- A page was added to interact with the ButtonDropdown component with the new item types.
- Unit tests were added or extended to test the new behavior and to test it doesn't change existing behavior. 

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_
- _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/button-dropdown/interfaces.ts [hunk 1]
```diff
@@ -10,37 +10,48 @@ import { ButtonProps } from '../button/interfaces';
...
+   * - `iconAlt` (string) - (Optional) Specifies alternate text for the icon when using `iconUrl`.
+   * - `iconUrl` (string) - (Optional) Specifies the URL of a custom icon.
+   * - `iconSvg` (ReactNode) - (Optional) Custom SVG icon. Equivalent to the `svg` slot of the [icon component](/components/icon/).
+   *
+   * ### group
+   *
+   * - `items` (ReadonlyArray<Item>): an array of item objects. Items will be rendered as nested menu items but only for the first nesting level, multi-nesting is not supported.
+   * An item which belongs to nested group has the following properties: `id`, `text`, `disabled` and `description`.
+   *
+   * ### radio-group
```
Comment (katiegeorge) [inline src/button-dropdown/interfaces.ts:50]: Radio group is out of scope right? We should remove this then.

File: src/button-dropdown/__tests__/button-dropdown-keyboard.test.tsx [hunk 2]
```diff
@@ -85,5 +86,47 @@ const items: ButtonDropdownProps.Items = [
...
+      // Open button dropdown again
+      act(() => wrapper.findNativeButton().keydown(KeyCode.enter));
+
+      // Fire keydown on the 1st element, checked should be undefined
+      act(() => wrapper.findItems()[0]!.keydown(KeyCode.enter));
+      expect(onClickSpy).toHaveBeenCalledTimes(2);
+      expect(onClickSpy).toHaveBeenCalledWith(expect.objectContaining({ detail: { id: 'i1', checked: undefined } }));
+    });
+
+    test('should fire event correctly when item with checkbox pressed using space', () => {
```
Comment (pan-kot) [inline src/button-dropdown/__tests__/button-dropdown-keyboard.test.tsx:110]: The tests are almost identical, you can reuse the code like so:
```
test.each([KeyCode.enter, KeyCode.space])('should fire event correctly when items with checkbox pressed using key=%s', (keyCode) => {
  // ...
});
```

File: src/button-dropdown/__tests__/button-dropdown.test.tsx [hunk 3]
```diff
@@ -317,3 +318,57 @@ describe('with main action', () => {
...
+      />
+    );
+  }
+
+  const { container } = render(<StateWrapper />);
+  const wrapper = createWrapper(container).findButtonDropdown()!;
+
+  // Click the item and verify it called the onClickSpy with the correct state and has correct value
+  wrapper.openDropdown();
+  expect(wrapper.findItemValueById('id')).toBe('true');
```
Comment (pan-kot) [inline src/button-dropdown/__tests__/button-dropdown.test.tsx:342]: Can we call the method `findItemCheckedById()` - this aligns with how the checked state is set in HTML. It took me a minute to get why the value is expected to be 'true'.

File: src/button-dropdown/item-element/index.tsx [hunk 4]
```diff
@@ -148,6 +162,22 @@ const MenuItemIcon = (props: InternalIconProps) => (
   </span>
 );
 
+// Toggle has aria-hidden set because it's just used as a graphical element,
+// a11y attributes for the checkmar are communicated through the role and aria-checked state
```
Comment (pan-kot) [inline src/button-dropdown/item-element/index.tsx:166]:
```suggestion
// a11y attributes for the checkmark are communicated through the role and aria-checked state
```

File: src/button-dropdown/utils/menu-item.ts [hunk 5]
```diff
@@ -20,3 +20,20 @@ export const getMenuItemProps: (params: GetMenuItemPropsParams) => React.HTMLPro
...
+export const getMenuItemCheckboxProps: (params: GetMenuItemCheckboxPropsParams) => React.HTMLProps<HTMLElement> = ({
+  disabled,
+  parent,
+  expanded,
+  checked,
+}: GetMenuItemCheckboxPropsParams) => ({
+  role: 'menuitemcheckbox',
+  'aria-disabled': disabled ? 'true' : undefined,
+  'aria-haspopup': parent ? 'true' : undefined,
+  'aria-expanded': expanded ? 'true' : parent ? 'false' : undefined,
```
Comment (pan-kot) [inline src/button-dropdown/utils/menu-item.ts:37]: why `aria-haspopup` and `aria-expanded` are needed on a checkbox item?

File: src/test-utils/dom/button-dropdown/index.ts [hunk 6]
```diff
@@ -38,6 +38,25 @@ export default class ButtonDropdownWrapper extends ComponentWrapper {
     return this.findOpenDropdown()?.find(itemSelector) || this.find(itemSelector);
   }
 
+  /**
+   * Finds `checked` value of item in the open dropdown by item id. Returns null if there is no open dropdown or the item is not a checkbox item.
+   *
+   * This utility does not open the dropdown. To find dropdown items, call `openDropdown()` first.
+   */
+  @usesDom findItemValueById(id: string): string | null {
+    console.log('yay');
```
Comment (pan-kot) [inline src/test-utils/dom/button-dropdown/index.ts:47]: yay!
```suggestion
```

File: src/button-dropdown/utils/menu-item.ts (hunk 5 above)
Comment (akoreman) [reply src/button-dropdown/utils/menu-item.ts:37]: You're right, those are not needed here, removed them.

File: src/button-dropdown/__tests__/button-dropdown.test.tsx (hunk 3 above)
Comment (akoreman) [reply src/button-dropdown/__tests__/button-dropdown.test.tsx:342]: yeah that makes sense, thanks for the suggestion.

**Comment by akoreman** [comment]
@pan-kot thanks for your comments, updated the PR to address them 🙂

**Review by pan-kot — DISMISSED** [review]

File: pages/button-dropdown/item-element.permutations.page.tsx [hunk 7]
```diff
@@ -149,7 +175,7 @@ export default function () {
         <PermutationsView
           permutations={permutations}
           render={permutation => (
-            <Box margin={{ bottom: 'xxl' }} className={styles['dropdown-permutation']}>
+            <Box margin={{ bottom: 'l' }} className={styles['dropdown-permutation']}>
```
Comment (pan-kot) [inline pages/button-dropdown/item-element.permutations.page.tsx:178]: Can we remove this change please? It makes the diff to permutations hard to review

Comment (akoreman) [reply pages/button-dropdown/item-element.permutations.page.tsx]: Hmm not sure why I even changed that, removed it.

**Review by pan-kot — DISMISSED** [review]

**Review by pan-kot — APPROVED** [review]

**Pull request merged**